# DEEP LEARNING
## PyTorch Basic

In [2]:
import torch

In [3]:
torch.cuda.is_available()

False

## Dataset 

In [14]:
import torch 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [15]:
df = pd.read_csv('Iris.csv')
df

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


In [16]:
df.shape

(150, 6)

In [20]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
le = LabelEncoder()
X = df.drop(['Species'], axis = 1).values
y = le.fit_transform(df['Species'].values)
#Chia du lieu voi test size = 20%

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)
X_train = torch.FloatTensor(X_train)
X_test = torch.FloatTensor(X_test)
y_train = torch.LongTensor(y_train).reshape(-1,1)
y_test = torch.LongTensor(y_test).reshape(-1,1)

In [19]:
len(y_train)

120

In [21]:
labels, count = y_train.unique(return_counts=True)
print(labels, count)

tensor([0, 1, 2]) tensor([40, 41, 39])


In [ ]:
# Tính đạo hàm bằng pytorch

# Cho y = 2x^4 + x^3 +3x^2 + 5x + 1 

# Tính y'

In [22]:
x = torch.tensor(2.0,requires_grad=True)
print(x)
print(x.grad)

tensor(2., requires_grad=True)
None


In [23]:
# Định nghĩa hàm
y = 2*x**4 + x**3 + 3*x**2 + 5*x + 1
print(y)

tensor(63., grad_fn=<AddBackward0>)


In [4]:
y.grad_fn

In [25]:
#Thực hiện lan truyền ngược và tính toán các gradient
y.backward()

In [26]:
#8x^3 + 3x^2 +6x +5
x.grad

tensor(93.)

In [ ]:
# BTVN 1
# Cho y = 5x^5 +6x^3 - 3x +1
# Cho biết độ dốc của đa thức trên ở điểm nào

In [10]:
import torch
# Khai báo biến x
x = torch.tensor(1.0, requires_grad=True)
# Định nghĩa hàm y
y = 5*x**5 + 6*x**3 - 3*x + 1
print(y)
y.grad_fn
# Tính đạo hàm 
y.backward()
print(x.grad)

tensor(9., grad_fn=<AddBackward0>)
tensor(40.)


In [ ]:
# BTVN 2
# Tạo một tensor ban đầu có giá trị là 2
# Định nghĩa hàm số và tính gradient
# y = x^3 + 2x^2 + 5x + 1
# Hãy tính dy/dx tại giá trị của x
# Dùng phương pháp Gradient Descent với Learning rate alpha = 0.1 để cập nhật giá trị x trong 10 vòng lặp

In [21]:
import torch

# tensor x = 2 và cho phép tính gradient
x = torch.tensor(2.0, requires_grad=True)

# Learning rate
alpha = 0.1

for i in range(10):
    # Định nghĩa hàm y
    y = x**3 + 2*x**2 + 5*x + 1

    # Tính gradient dy/dx
    y.backward()

    # In kết quả mỗi vòng
    print("Vòng:", i+1, ", x =", x.item(), ", dy/dx =", x.grad.item())

    # Gradient Descent: cập nhật x
    with torch.no_grad():
        x -= alpha * x.grad

    # Reset gradient về 0 để vòng sau tính lại
    x.grad.zero_()


Vòng: 1 , x = 2.0 , dy/dx = 25.0
Vòng: 2 , x = -0.5 , dy/dx = 3.75
Vòng: 3 , x = -0.875 , dy/dx = 3.796875
Vòng: 4 , x = -1.2546875476837158 , dy/dx = 4.703972339630127
Vòng: 5 , x = -1.7250847816467285 , dy/dx = 7.027413368225098
Vòng: 6 , x = -2.427826166152954 , dy/dx = 12.971714973449707
Vòng: 7 , x = -3.7249975204467773 , dy/dx = 31.726829528808594
Vòng: 8 , x = -6.897680282592773 , dy/dx = 120.14325714111328
Vòng: 9 , x = -18.912006378173828 , dy/dx = 1002.3439331054688
Vòng: 10 , x = -119.14640808105469 , dy/dx = 42116.01171875


In [ ]:
#BTVN
#Tạo một tập dữ liệu giả lập với x là số giờ học(ngẫu nhiên từ 1-10)và y là số điểm được tính 
#theo công thức y = 3x + 5 + nose với nose là một giá trị ngẫu nhiên nhỏ
#1. khởi tạo tham số w và b ngẫu nhiên với requires_grad = True
#2. Tính MSE
#3. Tính Gradient
#4. Cập nhật tham số w và b bằng Gradient Descent với Learning rate alpha = 0.1

In [92]:
import torch
x = torch.randint(1, 11, (20, 1)).float()  # 20 mẫu
noise = torch.randn(20, 1) # Tạo nhiễu ngẫu nhiên
y = 3 * x + 5 + noise
# Tạo tham số Pytorch tính gradient
w = torch.randn(1, requires_grad=True)
b = torch.randn(1, requires_grad=True)
# Mô hình hồi quy tuyến tính: y_pred = w*x + b
y_pred = w * x + b
# Hàm mất mát MSE (Mean Squared Error)
loss = torch.mean((y_pred - y) ** 2)
print(loss.item())
# Learning rate (tốc độ học)
alpha = 0.1
loss.backward()
print(loss)
# Cập nhật tham số
with torch.no_grad():
    w -= alpha * w.grad
    b -= alpha * b.grad
# Reset gradient
w.grad.zero_()
b.grad.zero_()

print(w)
print(b)


269.18328857421875
tensor(269.1833, grad_fn=<MeanBackward0>)
tensor([17.5695], requires_grad=True)
tensor([2.7494], requires_grad=True)


- 269.18328857421875 là giá trị được in bằng loss.item() dạng số thường(float) rất lớn vì w, b là ngẫu nhiên mới trải qua 1 bước học duy nhất.
- tensor(269.1833, grad_fn=<MeanBackward0>) loss được tính bằng hàm mean.
- tensor([17.5695], requires_grad=True) giá trị của w lớn do learning rate = 0.1
- tensor([2.7494], requires_grad=True) b nhỏ hơn w vì b chỉ là hằng số dịch chuyển, còn w nhân với x nên có ảnh hưởng lớn hơn và gradient lớn hơn

In [27]:
import torch 
import numpy as np

In [28]:
torch.__version__

'2.8.0+cpu'

In [29]:
#Chuyển đổi mảng numpy sang tensor pytorch
arr = np.array([1,2,3,4,5])
print(arr)
print(arr.dtype)
print(type(arr))

[1 2 3 4 5]
int64
<class 'numpy.ndarray'>


In [30]:
x = torch.from_numpy(arr)
print(x)

tensor([1, 2, 3, 4, 5])


In [31]:
print(x.dtype)

torch.int64


In [32]:
print(x.type())

torch.LongTensor


In [33]:
arr2 = np.arange(0.,12).reshape(4, 3)
arr2

array([[ 0.,  1.,  2.],
       [ 3.,  4.,  5.],
       [ 6.,  7.,  8.],
       [ 9., 10., 11.]])

In [34]:
x2 =  torch.from_numpy(arr2)
print(x2)
print(x2.type())

tensor([[ 0.,  1.,  2.],
        [ 3.,  4.,  5.],
        [ 6.,  7.,  8.],
        [ 9., 10., 11.]], dtype=torch.float64)
torch.DoubleTensor


In [35]:
#copy and sharing
x = torch.from_numpy(arr)
print(x)

tensor([1, 2, 3, 4, 5])


In [36]:
arr[0] = 99
print(x)

tensor([99,  2,  3,  4,  5])


In [37]:
arr = np.arange(0,5)
x = torch.tensor(arr)
print(x)

tensor([0, 1, 2, 3, 4])


In [38]:
arr[0] = 99
print(x)

tensor([0, 1, 2, 3, 4])


In [ ]:
#BTVN giải thích 2 trường hợp ở trên

- torch.tensor(arr)
Cho phép copy dữ liệu, không chia sẻ bộ nhớ nên thay đổi chuỗi nhưng không ảnh hưởng đến kết quả 
- torch.from_numpy(arr)
Không cho phép copy dữ liệu, chia sẻ dữ liệu chung nên thay đổi được chuỗi và làm ảnh hưởng đến kết quả.

In [ ]:
#BTVN 
# Bạn tạo giúp thầy tạo tensor với:
# empyty
# zeros
# ones
# random
# reshape với view và view as

In [ ]:
import torch

a = torch.empty(2, 3) #empyty

b = torch.zeros(2, 3) #zeros

c = torch.ones(2, 3) #ones

d = torch.rand(2, 3) #random

e = torch.arange(6)
e_view = e.view(2, 3) #reshape với view 

f = torch.rand(2, 3)
g = torch.arange(6)
g_view_as = g.view_as(f) #Reshape với view_as

print(a) #Tensor chưa khởi tạo giá trị
print(b) #Tensor toàn số 0
print(c) #Tensor toàn số 1
print(d) #Tensor ngẫu nhiên
print(e) #Tensor 1 chiều
print(e_view) #Tensor đã reshape bằng view
print(f) #Tensor ngẫu nhiên
print(g) #Tensor 1 chiều
print(g_view_as) #Tensor reshape theo shape của f


tensor([[-1.9673e-06,  1.3523e-42,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[0.9322, 0.0535, 0.0245],
        [0.8843, 0.7034, 0.8002]])
tensor([0, 1, 2, 3, 4, 5])
tensor([[0, 1, 2],
        [3, 4, 5]])
tensor([[0.7360, 0.6195, 0.3026],
        [0.6613, 0.9846, 0.9487]])
tensor([0, 1, 2, 3, 4, 5])
tensor([[0, 1, 2],
        [3, 4, 5]])
